perf-event/src/events.rs

```diff
@@ -318,3 +335,172 @@ pub enum CacheResult {
+        const READ = bindings::HW_BREAKPOINT_R;
+
+        /// Count when we write the memory location.
+        const WRITE = bindings::HW_BREAKPOINT_W;
+
+        /// Count when we read or write the memory location.
+        const READ_WRITE = Self::READ.union(Self::WRITE).bits();
+
+        /// Count when we execute code at the memory location.
+        const EXECUTE = bindings::HW_BREAKPOINT_X;
+    }
+}
+
+/// A hardware breakpoint.
+///
+/// A hardware breakpoint watches a region of memory for accesses. It has three
+/// parameters:
+/// - the address that is being watched (`addr`)
+/// - the number of bytes that breakpoint covers (`len`)
+/// - which type of memory accesses we care about (`ty`)
+///
+/// Note that both number of bytes that can be watched as well as the number of
+/// breakpoints that is allowed to be active at any given time is limited.
+///
+/// # Execute Breakpoint
+/// We can use a breakpoint to count the number of times that a function gets
+/// called, as long as the compiler does not optimize the function away.
+///
+/// ```
+/// # use perf_event::Builder;
+/// # use perf_event::events::Breakpoint;
+/// #[inline(never)]
+/// fn do_some_things() {
+///     // ...
+///     # println!("test println so the function doesn't get removed")
+/// }
+///
+/// let fnptr = do_some_things as fn() as usize;
+/// let mut counter = Builder::new()
+///     .kind(Breakpoint::execute(fnptr as u64))
+///     .build()?;
+/// counter.enable()?;
+///
+/// for _ in 0..500 {
+///     do_some_things();
+/// }
+///
+/// counter.disable()?;
+/// assert_eq!(counter.read()?, 500);
+/// # Ok::<(), std::io::Error>(())
+/// ```
+///
+/// # Data Breakpoint
+/// We can also use a breakpoint to count the number of times that a memory
+/// location is accessed.
+/// ```
+/// # use perf_event::Builder;
+/// # use perf_event::events::Breakpoint;
+/// #
+/// let mut data: Vec<u64> = (0..1024).rev().collect();
+///
+/// let mut counter = Builder::new()
+///     .kind(Breakpoint::read_write(&data[20] as *const _ as usize as u64, 8))
+///     .build()?;
+/// counter.enable()?;
+/// data.sort();
+/// counter.disable()?;
+///
+/// println!("Position 20 accessed {} times", counter.read()?);
+/// # Ok::<(), std::io::Error>(())
+/// ```
+///
+/// # Usage Notes
+/// - Some systems do not support creating read-only or write-only breakpoints.
+///   If you are getting `EINVAL` errors while trying to build such a counter
+///   using a read-write breakpoint might work instead.
+///
+/// - It is not valid to have a breakpoint that matches on both read/write and
+///   execute accesses. Trying to do this will result in an error.
+///
+/// - The valid values of len are quite limited. The [`perf_event_open`][man]
+///   manpage indicates that the only valid values for `bp_len` are 1, 2, 4,
+///   and 8.
+///
+/// [man]: http://man7.org/linux/man-pages/man2/perf_event_open.2.html
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Hash)]
+pub struct Breakpoint {
```

Comment:
I think we can tighten this up without making it any harder to use:

```
pub enum Breakpoint {
    Data { access: BreakpointAccess, addr: u64, len: u64 },
    Code { addr: u64 },
}

enum BreakpointAccess { Read, Write, ReadWrite }
```

This conveys the underlying API's constraints, but still seems straightforward to use. It could have an `update_attrs` method for `Event::update_attrs` to call.

We could use an enum for the length, too, but that just seems like going too far.

Reply:
That is nicer. I've gone and adjusted `Breakpoint` to match although I've left `BreakpointAccess` as a bitfield since it matches what is going on with the underlying data better.